pkg/api/operators/apis/experiment/v1alpha2/experiment_types.go

```diff
@@ -236,12 +236,64 @@ type Operation struct {
 	Parameters    []ParameterSpec `json:"parameters,omitempty"`
 }
 
-// Spec for metrics collectors. For v1alpha2 we will keep metrics collection as
-// cron jobs. This can be replaced by a push-model when the metadata client lib
-// is ready.
 type MetricsCollectorSpec struct {
-	Retain     bool       `json:"retain,omitempty"`
+	// Deprecated Retain
+	Retain bool `json:"retain,omitempty"`
+	// Deprecated GoTemplate
 	GoTemplate GoTemplate `json:"goTemplate,omitempty"`
+
+	Source    *SourceSpec    `json:"source,omitempty"`
+	Collector *CollectorSpec `json:"collector,omitempty"`
+}
+
+type SourceSpec struct {
+	// Model-train source code can expose metrics by http, such as HTTP endpoint in
+	// prometheus metric format
+	HttpGet *v1.HTTPGetAction `json:"httpGet,omitempty"`
+	// During training model, metrics may be persisted into local file in source
+	// code, such as tfEvent use case
+	FileSystemPath *FileSystemPath `json:"fileSystemPath,omitempty"`
+	// Default metric output format is {"metric": "<metric_name>",
+	// "value": <int_or_float>, "epoch": <int>, "step": <int>}, but if the output doesn't
+	// follow default format, please extend it here
+	Filter *FilterSpec `json:"filter,omitempty"`
```

Comment:
Are we doing a type check here when parsing?

Reply:
metricsCollector will try to convert the value to int or float and take it as an invalid entry with warning message if fail to convert. WDYT?

Reply:
Will this be a bigger constraint for the user?  There can be errors in Type cast of int vs float ?

Reply:
Currently, we are not doing any check when we write into DB https://github.com/kubeflow/katib/blob/master/pkg/api/v1alpha2/api.proto#L346

May be accept string and convert to float ?

Reply:
I just think that we can use `Filter` filed in term of extensibility.  If user source code metrics output style follows katib default suggested format, user doesn't need this field indeed. But if user source code doesn't follow our default style, we have no way to make it run by current katib except that user writes its new metricsCollector from scratch. in this case with this field, we at least can make it possible that user just point us which its format look like so that our metricsCollector can recognize a new pattern.
as for value type, I think maybe we can just guide users that we will always take it as float (for a db column, its value type can be uniquely defined)
In sum, `Filter` no harm but good IMO.

Reply:
Also, are we looking at multiple metrics in single line? (as in comment?)